lib/hmmlearn/vhmm.py

```diff
@@ -0,0 +1,626 @@
+import logging
+
+import numpy as np
+import numpy.linalg
+
+from scipy.special import digamma, logsumexp
+
+from sklearn import cluster
+from sklearn.utils import check_array, check_random_state
+
+from . import _hmmc, _utils
+from .base import ConvergenceMonitor, DECODER_ALGORITHMS, VariationalBaseHMM
+from .kl_divergence import kl_dirichlet, kl_multivariate_normal_distribution, \
+        kl_wishart_distribution
+from .stats import log_multivariate_normal_density
+from .utils import log_mask_zero, log_normalize, normalize
+
+
+_log = logging.getLogger(__name__)
+
+
+class VariationalCategoricalHMM(VariationalBaseHMM):
+    """
+    Hidden Markov Model with categorical (discrete) emissions
+    trained using Variational Inference.
+
+    References:
+        * https://cse.buffalo.edu/faculty/mbeal/thesis/
+
+    Attributes
+    ----------
+    n_features_ : int
+        Number of possible symbols emitted by the model (in the samples).
+
+    monitor_ : ConvergenceMonitor
+        Monitor object used to check the convergence of EM.
+
+    startprob_ : array, shape (n_components, )
+        Initial state occupation distribution.
+
+    transmat_ : array, shape (n_components, n_components)
+        Matrix of transition probabilities between states.
+
+    emissionprob_ : array, shape (n_components, n_features_)
+        Probability of emitting a given symbol when in each state.
+
+    Examples
+    --------
+    >>> from hmmlearn.hmm import VariationalCategoricalHMM
+    >>> VariationalCategoricalHMM(n_components=2)  #doctest: +ELLIPSIS
+    VariationalCategoricalHMM(algorithm='viterbi',...
+    """
+
+    def __init__(self, n_components=1,
+                 startprob_prior=None, transmat_prior=None,
+                 emissions_prior=None,
+                 algorithm="viterbi", random_state=None,
+                 n_iter=100, tol=1e-6, verbose=False,
+                 params="ste", init_params="ste",
+                 implementation="log"):
+        super().__init__(
+            n_components=n_components, startprob_prior=startprob_prior,
+            transmat_prior=transmat_prior,
+            algorithm=algorithm, random_state=random_state,
+            n_iter=n_iter, tol=tol, verbose=verbose,
+            params=params, init_params=init_params,
+            implementation=implementation
+        )
+
+        self.emissions_prior = emissions_prior
+
+    def _check_and_set_categorical_features(self, X):
```

Comment:
Perhaps introduce a `_BaseCategoricalHMM` that hosts the common methods (check_and_set_categorical_n_features, get_n_fit_scalars_per_param, etc.) and make both the nonvariational version and the variational one inherit from it?  This way it is easier to focus on just what's new for the variational version.

Reply:
If we can get the other PR merged to have a true MultinimialHMM, then I will rebase and try this.

Reply:
OK, let's try to get that other PR in first, I think it's quite close.

Reply:
Thank you for merging that other PR - I'll take a stab at this part next.

Reply:
@anntzer I think the way I've set things up a inheritance will be tricky as I understand it-

Here are the methods that are common:
   * `compute_likelihood`
   * `_generate_sample_from_state`
   * `_initialize_sufficient_statistics`
   * `_accumulate_sufficient_statistics`
   * ` _get_n_fit_scalars_per_param`
   * `_check_and_set_categorical_n_features` (already made free-standing)

Because of how inheritance is used, `_initialize_sufficient_statistics` and `_accumulate_sufficient_statistics`  would still need to call the parent method.  However, a `_BaseCategoricalHMM` would not derive from the necessary classes to still ensure that `super()._initialize_sufficient_statistics does the right thing.

I think we'd end up with something like this:
```mermaid
  graph TD;
      BaseEstimator-->BaseHMM
      BaseHMM-->VariationalBaseHMM;
      BaseHMM-->CategoricalHMM;
      VariationalBaseHMM-->VariationalCategoricalHMM;
      _BaseCategorial-HMM->CategoricalHMM;
      _BaseCategorial-HMM->VariationalCategoricalHMM;
```

So, I don't see a way to insert `_BaseCategoricalHMM` into the inheritance hierarchy and still make things work.  One option would be to at least make `_BaseCategoricalHMM`, with static methods, and have each of the Variational/EM classes call methods into it.  This would at least make the code centralized, and allow further evolution over time.

Reply:
Actually, super-calls should work, no?  I haven't actually looked at the details of which class implements what (so I may be misunderstanding your problem), but say you have
```
 Base
 /  \
B   C
 \ /
  D
```
and all 4 methods implement `foo` (i.e. `_initialize_sufficient_statistics`) with a super()-call (except for `Base`; this is important so that we don't end up trying calling the non-existent `object.foo()`):
```
class Base:
    def foo(self):
        print("base")
class A(Base):
    def foo(self):
        print("a"); return super().foo()
class B(Base):
    def foo(self):
        print("b"); return super().foo()
class C(A, B):
    def foo(self):
        print("c"); return super().foo()
```
then `C().foo()` will actually execute all four method bodies (printing `c a b base` -- note that the order of inheritance ends up determining whether A.foo or B.foo is called first).  (See e.g. http://python-history.blogspot.com/2010/06/method-resolution-order.html; also note that the old Py2 super-syntax, while longer, was clearer here: the result of the super-call in a method body actually depends on the exact type of `self`, not just what class we are currently in).

Or did I completely misunderstand your point?

Reply:
To move what I proposed, to this example, then I believe I would need `_BaseCategoricalHMM` to inherit from `BaseHMM`?,  We'll end up with two Diamond Patterns then, one for the `VariationalCategoricalHMM`, and one for the EM `CategoricalHMM`:

```mermaid
  graph TD;
      BaseEstimator-->BaseHMM;
      BaseHMM-->VariationalBaseHMM;
      BaseHMM-->CategoricalHMM;
      VariationalBaseHMM-->VariationalCategoricalHMM;
     BaseEstimator -->_BaseCategorialHMM;
      _BaseCategorialHMM-->CategoricalHMM;
      _BaseCategorialHMM-->VariationalCategoricalHMM;
```

Reply:
@alexrockhill @anntzer 

Making Emissions Specific classes greatly simplified things.  Right now I stuck them in `base.py`, but should we create `emissions.py` (or something similar?)  This should hopefully make the review easier.

Also, apologies for how long this has taken!

Reply:
Sure, moving base emissions to emissions.py seems reasonable.
No worries for the time, we all have many other priorities :)

Reply:
We've since restructured everything.